internal/config/instance.go

```diff
@@ -119,6 +119,22 @@ func (i *Instance) setWithCallback(key string, valueF func(currentValue interfac
 		return nil
 	}
 
+	// Cast to rule type if applicable
+	rule := GetRule(key)
+	switch rule.Type {
+	case Bool:
+		v = cast.ToBool(v)
+	case Int:
+		v = cast.ToInt(v)
+	case String:
+		v = cast.ToString(v)
+	}
+
+	err = rule.SetEvent(v)
```

Comment:
Can we make this more explicit by returning the intended new value of `v` rather than modifying the pointer?